From e43f81579a6d4bf2d2ac6a378e85ddfb42b38870 Mon Sep 17 00:00:00 2001
From: Juan Gutierrez <juan.gutierrez@nxp.com>
Date: Tue, 31 Jan 2017 10:32:19 -0600
Subject: [PATCH 1491/5242] MXSCM-243-1 arm: imx: get the mmdc 2ch-mode api
 for lpddr2

commit  5ebd31c3d158d0a02409e5f541e09f750f01d1b1 from
https://source.codeaurora.org/external/imx/linux-imx.git

To configure the suspend settings for lpddr2 systems is necessary
to know if mmdc is operating on 1ch-mode or 2ch-mode.
Here, the imx_get_lpddr2_2ch_mode api is introduced to get this info
when needed and decide accordingly.

Signed-off-by: Juan Gutierrez <juan.gutierrez@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-imx/common.h |    2 ++
 arch/arm/mach-imx/mmdc.c   |   11 +++++++++++
 arch/arm/mach-imx/mxc.h    |    3 +++
 3 files changed, 16 insertions(+)

diff --git a/arch/arm/mach-imx/common.h b/arch/arm/mach-imx/common.h
index 697cc9c8ad..28ab41a 100644
--- a/arch/arm/mach-imx/common.h
+++ b/arch/arm/mach-imx/common.h
@@ -160,8 +160,10 @@ static void imx_gpcv2_add_m4_wake_up_irq(u32 hwirq, bool enable) {}
 int imx7ulp_set_lpm(enum imx7ulp_cpu_pwr_mode mode);
 #ifdef CONFIG_HAVE_IMX_MMDC
 int imx_mmdc_get_ddr_type(void);
+int imx_mmdc_get_lpddr2_2ch_mode(void);
 #else
 static inline int imx_mmdc_get_ddr_type(void) { return 0; }
+static inline int imx_mmdc_get_lpddr2_2ch_mode(void) { return 0; }
 #endif
 #ifdef CONFIG_HAVE_IMX_DDRC
 int imx_ddrc_get_ddr_type(void);
diff --git a/arch/arm/mach-imx/mmdc.c b/arch/arm/mach-imx/mmdc.c
index 04b3bf7..1cc977f 100644
--- a/arch/arm/mach-imx/mmdc.c
+++ b/arch/arm/mach-imx/mmdc.c
@@ -31,6 +31,8 @@
 #define MMDC_MDMISC		0x18
 #define BM_MMDC_MDMISC_DDR_TYPE	0x18
 #define BP_MMDC_MDMISC_DDR_TYPE	0x3
+#define BM_MMDC_MDMISC_LPDDR2_2CH	0x4
+#define BP_MMDC_MDMISC_LPDDR2_2CH	0x2
 
 #define TOTAL_CYCLES		0x0
 #define BUSY_CYCLES		0x1
@@ -64,6 +66,7 @@
 #define to_mmdc_pmu(p) container_of(p, struct mmdc_pmu, pmu)
 
 static int ddr_type;
+static int lpddr2_2ch_mode;
 
 struct fsl_mmdc_devtype_data {
 	unsigned int flags;
@@ -556,6 +559,9 @@ static int imx_mmdc_probe(struct platform_device *pdev)
 	val = readl_relaxed(reg);
 	ddr_type = (val & BM_MMDC_MDMISC_DDR_TYPE) >>
 		 BP_MMDC_MDMISC_DDR_TYPE;
+	/* Get lpddr2 2ch-mode */
+	lpddr2_2ch_mode = (val & BM_MMDC_MDMISC_LPDDR2_2CH) >>
+			BP_MMDC_MDMISC_LPDDR2_2CH;
 
 	reg = mmdc_base + MMDC_MAPSR;
 
@@ -572,6 +578,11 @@ int imx_mmdc_get_ddr_type(void)
 	return ddr_type;
 }
 
+int imx_mmdc_get_lpddr2_2ch_mode(void)
+{
+	return lpddr2_2ch_mode;
+}
+
 static struct platform_driver imx_mmdc_driver = {
 	.driver		= {
 		.name	= "imx-mmdc",
diff --git a/arch/arm/mach-imx/mxc.h b/arch/arm/mach-imx/mxc.h
index 5dcda87..f162e82 100644
--- a/arch/arm/mach-imx/mxc.h
+++ b/arch/arm/mach-imx/mxc.h
@@ -49,6 +49,9 @@
 #define IMX_DDR_TYPE_LPDDR3		2
 #define IMX_MMDC_DDR_TYPE_LPDDR3	3
 
+#define IMX_LPDDR2_1CH_MODE            0
+#define IMX_LPDDR2_2CH_MODE            1
+
 #ifndef __ASSEMBLY__
 extern unsigned int __mxc_cpu_type;
 
-- 
1.7.9.5

